### PR TITLE
Initial contribution of ZigBee binding

### PR DESCRIPTION
**Pull request by cdjackson**
Currently this is functioning with Hue lights. It should work with any ZigBee compatible (ZLL/HA) light, but I've only tested on Hue. This supports generation of dynamic channels from the list of clusters supported by the device and I'll add to the supported clusters shortly, and will also add configuration options, but I thought it would be good to get something out for review.

Signed-off-by: Chris Jackson chris@cd-jackson.com

### CODE REVIEW

**Comment by cdjackson** [comment]
Also tested with Ubisys-S2 dual switch...

**Comment by ideologiefrei** [comment]
I'd like to test your zigbee binding, The TI USB hardware is already on my desk. Any Plans on working on this PR, or should i try to compile it myself? ;)

**Comment by cdjackson** [comment]
@kaikreuzer I wonder if we can look to merge this while it's actively developed - in the same way as the zwave binding?

I think the error reported above was a test failure in another binding (Max I think).

**Comment by kaikreuzer** [comment]
Hey @cdjackson I didn't forget about you! I am currently thinking about adding a specific Karaf feature for  "experimental" add-ons - the ZigBee binding would be a candiate for that. I am a bit undecided yet, whether this should all go directly to the master branch or maybe work on it on a separate branch instead; in any case, it should fulfill from the start the requirements listed in https://www.eclipse.org/smarthome/documentation/development/guidelines.html.

File: addons/binding/org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/cluster/ZigBeeClusterHandler.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    public static ZigBeeClusterHandler getConverter(int clusterId) {
+        if (clusterMap == null) {
+            clusterMap = new HashMap<Integer, Class<? extends ZigBeeClusterHandler>>();
+
+            // Add all the handlers into the map...
+            clusterMap.put(ZigBeeApiConstants.CLUSTER_ID_ON_OFF, ZigBeeOnOffClusterHandler.class);
+            clusterMap.put(ZigBeeApiConstants.CLUSTER_ID_LEVEL_CONTROL, ZigBeeLevelClusterHandler.class);
+            clusterMap.put(ZigBeeApiConstants.CLUSTER_ID_COLOR_CONTROL, ZigBeeColorClusterHandler.class);
+            clusterMap.put(ZigBeeApiConstants.CLUSTER_ID_RELATIVE_HUMIDITY_MEASUREMENT,
+                    ZigBeeRelativeHumidityMeasurementClusterHandler.class);
```
Comment (marcelrv) [inline addons/binding/org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/cluster/ZigBeeClusterHandler.java]: @cdjackson Seems that ZigBeeRelativeHumidityMeasurementClusterHandler.class
 & ZigBeeTemperatureMeasurementClusterHandler.class are missing in your PR

Comment (cdjackson) [inline addons/binding/org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/cluster/ZigBeeClusterHandler.java]: Ok - I’ll update this to remove the references - I didn’t mean to add them yet…

**Comment by tavalin** [comment]
@kaikreuzer @cdjackson now that we have an experimental repo, is this something we can progress?

**Comment by kaikreuzer** [comment]
Almost, but we would also like to move it to a separate repository, similar to what we plan with Z-Wave. So please allow us just a little bit more time :-)

**Comment by kgoderis** [comment]
> Almost, but we would also like to move it to a separate repository, similar to what we plan with Z-Wave. So please allow us just a little bit more time :-)

@kaikreuzer Kai, is there a place where we can discuss this, or is discussed already?

**Comment by tavalin** [comment]
Has there been any movement on this?

**Comment by bokas74** [comment]
Excellent initiative! Would love to see this binding as part of the master. Waiting to use this contrib from Cris, once again!

**Comment by Arsur** [comment]
Hi
This binding should also work with Hue Dimming switch and Osram Bulbs isnt it?
I orderd one of this dongles from china but it will take a month delivery time i think until i can test it :/

**Comment by cdjackson** [comment]
Yes - it should work fine.

**Comment by tstasevych** [comment]
Guys, what is the status of this? Will this ever be as a experimental binding, or we need to compile ourselves? Anyone has jar available ??

**Comment by PaulKC** [comment]
I used the firmware described here:
https://github.com/tlaukkan/zigbee4java

CC2531ZNP-Pro-Secure_Standard.hex

**Comment by jfreedman** [comment]
@Arsur, did your Chinese dongle work?  Way cheaper than the $50 TI one they want you to get.

**Comment by Arsur** [comment]
> @Arsur, did your Chinese dongle work? Way cheaper than the $50 TI one they want you to get.

i had no time to try it yet. But i have now holidays...

**Comment by derabbink** [comment]
I created https://github.com/cdjackson/openhab2-addons/pull/16 to fix the merge conflict. @cdjackson could update this PR using that, if they like.
Other than that tiny change to `addons/binding/pom.xml`, is there anything else missing before this can make it to master?

**Comment by Arsur** [comment]
Good news i think i made some progress. I got the Chinese - CC Debugger and - Zigbee Stick working. I flashed CC2531ZNP-Pro-Secure_Standard.hex firmware on it. I think now i have to wait for the binding now isnt it?
Here are the links for the aliexpress products i successfully combined
[Cable](https://www.aliexpress.com/item/Downloader-Cable-Bluetooth-4-0-CC2540-zigbee-CC2531-Sniffer-USB-dongle-BTool-Programmer-Wire-Download-Programming/32767478130.html)
[Zigbee CC2531 Stick](https://www.aliexpress.com/item/Wireless-Zigbee-CC2531-Sniffer-Packet-Protocol-Analyzer-Module-USB-Interface-Dongle-Capture-Packet/32769663830.html)
[CC Debugger](https://www.aliexpress.com/item/CC-Debugger-And-Programmer-Downloader-For-RF-System-On-Chips-ICSH015A/32672493016.html)

**Comment by giannello** [comment]
Quick update on the zigbee topic: I've ordered the products mentioned by @Arsur and used https://github.com/dashesy/cc-tool to flash https://github.com/ninjablocks/node-zigbee/blob/master/firmware/CC2531ZNP-Pro-Secure_Standard.hex using Linux.
To make the software detect the debugger AND the dongle, connect them together and before plugging the debugger to the pc, keep the 'debug' button pressed on the dongle.

**Comment by DuncanvR** [comment]
Like @giannello, I also got the hardware @Arsur mentioned, flashed the firmware and played around with it yesterday and today (only took me about a year 🙈).

I used the `CC2531ZNP-Pro-Secure_Standard.hex` file that came with TI's Smart RF Studio, as described in the [zigbee4java](https://github.com/tlaukkan/zigbee4java) readme. Haven't tried the ones linked to by @giannello.

After lots of fiddling about, I got the zigbee addon from this PR working in openHAB 2. Initially I didn't use Eclipse Smarthome, so I had to copy dependencies and modify the `build.properties` and `MANIFEST.MF` files to reference those. I also found I needed the `RXTXcomm.jar` library, which comes with the [java-rxtx](https://www.archlinux.org/packages/community/x86_64/java-rxtx) package in Arch.

At first the binding wouldn't find my device, regardless of the port I specified. Since going through openHAB to test it was a bit cumbersome, I used the following code to test the connection. Add the `RXTXcomm.jar` to your classpath in order to run it.
```
import gnu.io.CommPortIdentifier;

try {
    System.out.println("Available ports:");
    java.util.Enumeration ports = CommPortIdentifier.getPortIdentifiers();
    while (ports.hasMoreElements()) {
        CommPortIdentifier port = (CommPortIdentifier) ports.nextElement();
        System.out.println("\t" + port.getName());
    }
    System.out.println("\t</>");
} catch (Exception e) {
   System.out.println("\tUnable to list available ports: " + e.getMessage());
}
```

I first had to update my kernel to have the dongle listed under `/dev`. Secondly there were issues with write permissions to `/run/lock`, which the serial library uses. The Arch package actually has a patch to use the `/run/lock/lockdev` folder for those locks and a post install message stating what needs to be done:
> Users need to be in 'lock' and 'uucp' groups to connect to devices


After all this the addon was picking up my dongle and going through the initialisation phase successfully. Unfortunately it doesn't detect my Lightify bulb, so that's where I'm stranded at the moment.

Judging from the activity at https://github.com/openhab/org.openhab.binding.zigbee however, I suspect @cdjackson will be closing this PR soon and development of this addon will be moved to that repo. I'll be submitting a PR there with some minor fixes I made this afternoon and look forward to working on this further!

**Comment by kaikreuzer** [comment]
The code has moved to https://github.com/openhab/org.openhab.binding.zigbee, so I am closing this PR here.

**Pull request closed without merging**